string.c

```diff
@@ -1480,6 +1480,14 @@ rb_str_new_frozen_String(VALUE orig)
     return str_new_frozen(rb_cString, orig);
 }
 
+
+VALUE
+rb_str_frozen_bare_string(VALUE orig)
+{
+    if (RB_LIKELY(BARE_STRING_P(orig))) return orig;
```

Comment:
Hum, that doesn't do what's marked on the tin. Here you are returning a string that is potentially mutable.

Reply:
Oh yeah thanks for the catch, updated that function and added two tests, one checking that an non-frozen string turns into a (new) frozen one, and one checking that a frozen String with no instance variables just gets set.